DependencyInjection/CompilerPass/FixturesCompilerPass.php

```diff
@@ -21,9 +22,30 @@ public function process(ContainerBuilder $container)
 
         $fixtures = [];
         foreach ($taggedServices as $serviceId => $tags) {
-            $fixtures[] = new Reference($serviceId);
+            $fixtures[] = [
+                'fixture' => new Reference($serviceId),
+                'groups' => $this->getFixtureGroups($serviceId, $container)
+            ];
+
         }
 
         $definition->addMethodCall('addFixtures', [$fixtures]);
     }
+
+    private function getFixtureGroups($service, ContainerBuilder $container)
```

Comment:
Would it not make sense to defer this work of figuring out the groups to the SymfonyFixturesLoader?

Here you do have a lot of logic to figure out which class implements each fixture. It complicates the code, and it could also fail if a fixture service was defined by a factory method. ($def->getClass() can be different from the actual class in that case).

In the SymfonyFixturesLoader all the fixtures has been resolved to actual objects, and you could just use that object to figure out if it implemets the interface and call "getGroups" on it.

Reply:
I'm conflicted. On one hand, you're right - the service instance might be a different class than the one specified in the container. On the other hand, moving it to the fixture loader would require instantiating every fixture in order to check if it's relevant, which can be avoided this way.

I'm not sure how relevant the use-case mentioned here is: could/should the Container complain if the instance a factory returns is of a different class than specified on the service?

Reply:
Instatiation of every fixture service should not be that much overhead. We are talking microseconds here in code that not very likely to be time critical any way.

I think it is much more important that the code always does the right thing.

Reply:
Actually it is very common that a factory returns a subclass of the one declared in the container. And some subclasses could implement the FixtureGroupInterface while others do not. I think that could be a very likely scenario.

Reply:
There's a long-standing pattern where these types of things are static - e.g. command class names, or `EventSubscriberInterface::getSusbcribedEvents()`. It makes things lazy, and instantiating objects that the user never asked to be used (e.g. fixtures in other groups) will probably lead to some WTF moments.

But, no problem! Another long-standing pattern is to add a second, non-static way of doing things: tags (this is true for both commands and event listeners). We'll do the same thing - allowing you to use that method if you need to.

Reply:
Since the SymfonyFixtureLoader gets an instance of all the fixtures anyway, (in the method addFixtures), no overhead is added by deferring this work to that method. And then it can be done right.

Please see https://github.com/weaverryan/DoctrineFixturesBundle/pull/1